server/src/main/java/org/opensearch/action/support/master/MasterThrottlingRetryListener.java

```diff
@@ -0,0 +1,135 @@
+import org.opensearch.action.bulk.BackoffPolicy;
+import org.opensearch.cluster.metadata.ProcessClusterEventTimeoutException;
+import org.opensearch.cluster.service.MasterTaskThrottlingException;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.threadpool.Scheduler;
+import org.opensearch.transport.TransportException;
+
+import java.util.Iterator;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * ActionListener for retrying the Throttled master tasks.
+ * It schedules the retry on the Throttling Exception from master node and
+ * delegates the response if it receive response from master.
+ *
+ * It uses ExponentialEqualJitterBackoff policy for determining delay between retries.
+ */
+public class MasterThrottlingRetryListener<Request extends MasterNodeRequest<Request>, Response extends ActionResponse>
+        implements ActionListener<Response>  {
+
+    private static final Logger logger = LogManager.getLogger(MasterThrottlingRetryListener.class);
+
+    /**
+     * Base delay in millis.
+     */
+    private final int BASE_DELAY_MILLIS = 10;
+
+    /**
+     * Maximum delay in millis.
+     */
+    private final int MAX_DELAY_MILLIS = 5000;
+
+    private long totalDelay;
+    private final Iterator<TimeValue> backoffDelay;
+    private final ActionListener<Response> listener;
+    private final Request request;
+    private final Runnable runnable;
+    private final String actionName;
+    private final boolean localNodeRequest;
+
+    private static ScheduledThreadPoolExecutor scheduler = Scheduler.initScheduler(Settings.EMPTY);
+
+    public MasterThrottlingRetryListener(
+            String actionName,
+            Request request,
+            Runnable runnable,
+            ActionListener<Response> actionListener) {
+        this.actionName = actionName;
+        this.listener = actionListener;
+        this.request = request;
+        this.runnable = runnable;
+        this.backoffDelay = BackoffPolicy.exponentialEqualJitterBackoff(BASE_DELAY_MILLIS, MAX_DELAY_MILLIS).iterator();
+        /**
+         This is to determine whether request is generated from local node or from remote node.
+         If it is local node's request we need to perform the retries on this node.
+         If it is remote node's request, we will not perform retries on this node and let remote node perform the retries.
+
+         If request is from remote data node, then data node will set remoteRequest flag in {@link MasterNodeRequest}
+         and send request to master, using that on master node we can determine if the request was localRequest or remoteRequest.
+         */
+        this.localNodeRequest = !(request.isRemoteRequest());
+    }
+
+    @Override
+    public void onResponse(Response response) {
+        listener.onResponse(response);
+    }
+
+    @Override
+    public void onFailure(Exception e) {
```

Comment:
Please add BWC tests for the following cases:
1. Old data node, new master node: Here, master may throttle the request and data node will propagate the throttling error to client, which may not be gracefully handled upstream. One way to get around this problem is to make master node know which node the request came from and handle the requests accordingly- e.g add a field in `MasterNodeRequest` about the same. Do you plan to handle this in a different way?
2. Old master node, new data node: This will work, I think. I think the setting update itself will fail.